// @dfuse/client/dist/lib/message/inbound.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// **Important** The key must be the same as the API type but in upper snake case for "in" operation to work
var InboundMessageType;
(function (InboundMessageType) {
    InboundMessageType["ACTION_TRACE"] = "action_trace";
    InboundMessageType["ERROR"] = "error";
    InboundMessageType["LISTENING"] = "listening";
    InboundMessageType["HEAD_INFO"] = "head_info";
    InboundMessageType["PING"] = "ping";
    InboundMessageType["PROGRESS"] = "progress";
    InboundMessageType["UNLISTENED"] = "unlistened";
    InboundMessageType["TABLE_DELTA"] = "table_delta";
    InboundMessageType["TABLE_SNAPSHOT"] = "table_snapshot";
    InboundMessageType["TRANSACTION_LIFECYCLE"] = "transaction_lifecycle";
})(InboundMessageType = exports.InboundMessageType || (exports.InboundMessageType = {}));
// export type InboundMessageType = "action_trace"
//                                | "error"
//                                | "listening"
//                                | "head_info"
//                                | "ping"
//                                | "progress"
//                                | "unlistened"
//                                | "table_delta"
//                                | "table_snapshot"
//                                | "transaction_lifecycle"
